core/src/main/java/io/grpc/Status.java

```diff
@@ -316,12 +320,53 @@ public static Status fromCodeValue(int codeValue) {
   public static final Metadata.Key<Status> CODE_KEY
       = Metadata.Key.of("grpc-status", new StatusCodeMarshaller());
 
+  private static final AsciiMarshaller<String> STATUS_MESSAGE_MARSHALLER =
+      new AsciiMarshaller<String>() {
+
+    @Override
+    public String toAsciiString(String value) {
+      // This can be made faster if necessary.
+      StringBuilder sb = new StringBuilder(value.length());
+      for (byte b : value.getBytes(Charset.forName("UTF-8"))) {
+        if (b >= ' ' && b < '%' || b > '%' && b < '~') {
+          // fast path, if it's plain ascii and not a percent, pass it through.
+          sb.append((char) b);
+        } else {
+          sb.append(String.format("%%%02X", b));
+        }
+      }
+      return sb.toString();
+    }
+
+    @Override
+    public String parseAsciiString(String value) {
+      Charset transerEncoding = Charset.forName("US-ASCII");
+      // This can be made faster if necessary.
+      byte[] source = value.getBytes(transerEncoding);
+      ByteBuffer buf = ByteBuffer.allocate(source.length);
+      for (int i = 0; i < source.length; ) {
+        if (source[i] == '%' && i + 2 < source.length) {
+          try {
+            buf.put((byte)Integer.parseInt(new String(source, i + 1, 2, transerEncoding), 16));
+            i += 3;
+            continue;
+          } catch (NumberFormatException e) {
+            // ignore, fall through, just push the bytes.
+          }
+        }
+        buf.put(source[i]);
+        i += 1;
+      }
+      return new String(buf.array(), 0, buf.position(), Charset.forName("UTF-8"));
```

Comment:
In this case UTF-8 is correct.  transfer encoding is the encoding of the header value bytes (sorry for the poor name, I'm running out!)

In this case, I want to treat the header value bytes as ascii, convert them into utf8 bytes, and then reparse _that_ byte array as a utf-8 string.  

This probably warrants better documentation.